designs/0100-reduce_sum.md

```diff
@@ -0,0 +1,214 @@
+
+reduce_sum is not useful, however, when there are dependencies between these likelihood terms. This can happen, for instance, if there were N terms in a Gaussian process likelihood. reduce_sum will not be useful for accelerating this.
+
+The signature we settled on for reduce_sum (which balances the different tradeoffs described in the Summary) is:
+
+```
+real reduce_sum(F func, T[] sliced_arg, int grainsize, T1 arg1, T2 arg2, ...)
+```
+
+where the function ```func``` has a signature like:
+
+```
+real func(int start, int end, T[] subset_sliced_arg, T1 arg1, T2 arg2, ...)
+```
+
+The array ```sliced_arg``` in the reduce_sum call represents the list of calculations to perform and sum.
+
+The user-provided function ```func``` is expect to compute the ```start``` through ```end``` terms of the overall sum, accumulate them, and return that value. The user function is only passed the subset ```sliced_arg[start:end]``` of sliced arg (as ```subset_sliced_arg```). ```start``` and ```end``` are passed so that ```func``` can index any of the ```argN``` appropriately. The trailing arguments ```argN``` are passed without modification to every call of ```func```.
+
+An overall call to ```reduce_sum``` can be replaced by either:
+
+```
+real sum = func(1, size(sliced_arg), sliced_arg, arg1, arg2, ...)
+```
+
+or (modulo differences due to rearrangements of summations) the code:
+
+```
+real sum = 0.0;
+for(i in 1:size(sliced_arg)) {
+  sum = sum + func(i, i, { sliced_arg[i] }, arg1, arg2, ...);
+}
+```
+
+The argument ```sliced_arg``` is called the sliced argument because func only receives the elements of ```sliced_arg``` for which it is responsible (in ```subset_sliced_arg```). All the other ```argN``` arguments are shared completely between every call to func.
+
+By requiring that the user provided function perform the calculation over a range of inputs, it is possible for the scheduler to lump smaller pieces of work together into big ones (and increase efficiency). This is done by adjusting the ```grainsize``` argument. The scheduler will try to cut up the N parallel terms into groups of size grainsize.
+
+As an example, the hierarchical model code:
+```
+vector[N_subset] mu = temp_Intercept + Xc[1:N_subset] * b;
+for (n in 1:N_subset) {
+  mu[n] += r_1_1[J_1[n]] * Z_1_1[n] + r_1_2[J_1[n]] * Z_1_2[n] +
+    r_2_1[J_2[n]] * Z_2_1[n] +
+    r_3_1[J_3[n]] * Z_3_1[n] + r_3_2[J_3[n]] * Z_3_2[n] +
+    r_4_1[J_4[n]] * Z_4_1[n] +
+    r_5_1[J_5[n]] * Z_5_1[n] +
+    r_6_1[J_6[n]] * Z_6_1[n] + r_6_2[J_6[n]] * Z_6_2[n] +
+    r_7_1[J_7[n]] * Z_7_1[n] + r_7_2[J_7[n]] * Z_7_2[n] +
+    r_8_1[J_8[n]] * Z_8_1[n] +
+    r_9_1[J_9[n]] * Z_9_1[n] +
+    r_10_1[J_10[n]] * Z_10_1[n] +
+    r_11_1[J_11[n]] * Z_11_1[n] + r_11_2[J_11[n]] * Z_11_2[n];
+}
+target += binomial_logit_lpmf(Y[1:N_subset] | trials[1:N_subset], mu);
+```
+
```

Comment:
why such a complicated example?  to illustrate a complex shared calculation?

Reply:
Yes, and that was easy to grab. We'll do a simpler one for the docs.